### PR TITLE
Have OpenMMParameterSet remove non-chemical H-H bonds from water upon creation

### PR DESCRIPTION
**Pull request by jchodera**
This PR partially addresses https://github.com/ParmEd/ParmEd/issues/941

- [x] `OpenMMParameterSet.from_parameterset()` now removes unphysical H-H bonds from waters
- [x] Added an `empirical_chemical_formula` property to `Residue` that returns the Hill notation chemical formula (e.g. `H2O`, `C6H6`), provided the `Atom` objects in the residue have `atomic_number` correctly set. Note that the CHARMM parameter reader does not actually set the `atomic_number` fields, so this must be done defensively by `OpenMMParameterSet.from_parameterset()`. (If it's a bug that the CHARMM reader doesn't set this correctly, we should fix that.)
- [x] Added `delete_bond` method to `Residue`
- [x] Added `element_name` field to `Atom`

Currently, the [`test_write_xml_parameters_charmm`](https://github.com/ParmEd/ParmEd/blob/master/test/test_parmed_openmm.py#L612-L633) will fail because, once H-H bonds are removed, there are now two identical water residue templates in [`toppar_water_ions.str`](https://github.com/ParmEd/ParmEd/blob/master/test/files/toppar_water_ions.str#L65-L87), causing OpenMM's `ForceField` to throw an exception. This means that there is no way for ParmEd to correctly read CHARMM parameter sets and write them to OpenMM if they contain duplicate waters, as the standard `toppar_water_ions.str` does. 

To fix this, I am currently presuming we will need to add the capability to specify which residue templates to omit when reading a CHARMM parameter file (pending response to https://github.com/choderalab/openmm-forcefields/issues/56#issuecomment-359291070).

cc: @peastman
cc: https://github.com/choderalab/openmm-forcefields/issues/56
cc: https://github.com/pandegroup/openmm/issues/1939#issuecomment-359130400

### CODE REVIEW

**Comment by jchodera** [comment]
@swails: I could use your guidance on how best to allow us to specify a sensible behavior for directing CHARMM -> OpenMM conversions to exclude certain topology residues to avoid duplicates. Some ideas:
1. Add an initializer kwarg:
```python
        params = CharmmParameterSet(*charmm_files, exclude_residue_templates=['TP3M'])
```
2. Simply use `del` to manually delete problematic residues in tests
```python
# read CHARMM parameters
charmm_params = pmd.charmm.CharmmParameterSet(get_fn('par_all36_prot.prm'),
                                              get_fn('top_all36_prot.rtf'),
                                              get_fn('toppar_water_ions.str'))
# drop the problematic residue to avoid the next step throwing an exception
del charmm_params.residues['TP3M']
# convert to OpenMM
openmm_params = openmm.OpenMMParameterSet.from_parameterset()
```
3. Something else?

**Comment by swails** [comment]
For now I’d say just use del to remove offending residue templates. 

But why not have a separate set of xml files for the different water models?

**Comment by jchodera** [comment]
That's the plan! I just wanted to figure out if we wanted to add additional API support to ParmEd to make it possible to read CHARMM parameter sets that could be exported without further manipulation. I'll use the del approach for now to fix the test and then forge on with the CHARMM conversion scripts.

**Comment by swails** [comment]
Honestly to me the question boils down to whether it makes sense to run the del call inside the function by specifying what to del as an argument to the reader vs. just making the client run the del command themselves. I’d rather not add that complexity to ParmEd.

**Comment by jchodera** [comment]
Hm:
```
======================================================================
ERROR: Tests automatic loading of downloaded Amber mdcrd file
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ParmEd/ParmEd/test/test_parmed_formats.py", line 2238, in test_download_amber_traj_ascii
    self.assertTrue(amber.AmberMdcrd.id_format(self.url + 'tz2.truncoct.crd'))
  File "/home/travis/.pyenv/versions/pypy-4.0.1/site-packages/parmed/amber/asciicrd.py", line 485, in id_format
    f = genopen(filename, 'r')
  File "/home/travis/.pyenv/versions/pypy-4.0.1/site-packages/parmed/utils/io.py", line 71, in genopen
    raise IOError('Could not open %s: %s' % (name, e))
IOError: Could not open https://github.com/ParmEd/ParmEd/raw/master/test/files/tz2.truncoct.crd: HTTP Error 504: Gateway Time-out
```

**Comment by jchodera** [comment]
Whoa. No idea what this means:
```
======================================================================
ERROR: Test NetCDF and ASCII restart and trajectory reporters (no PBC)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ParmEd/ParmEd/test/test_openmm_reporters.py", line 194, in test_reporters
    ntraj = [NetCDFTraj.open_old(get_fn('traj1.nc', written=True)),
  File "/home/travis/miniconda/envs/myenv/lib/python2.7/site-packages/parmed/amber/netcdffiles.py", line 544, in open_old
    inst = cls(fname, 'r')
  File "/home/travis/miniconda/envs/myenv/lib/python2.7/site-packages/parmed/amber/netcdffiles.py", line 406, in __init__
    self._ncfile = NetCDFFile(fname, mode, mmap=False)
  File "/home/travis/miniconda/envs/myenv/lib/python2.7/site-packages/parmed/utils/netcdf.py", line 286, in __init__
    self._read()
  File "/home/travis/miniconda/envs/myenv/lib/python2.7/site-packages/parmed/utils/netcdf.py", line 598, in _read
    self.__dict__['version_byte'] = fromstring(self.fp.read(1), '>b')[0]
DeprecationWarning: The binary mode of fromstring is deprecated, as it behaves surprisingly on unicode inputs. Use frombuffer instead
```

**Comment by jchodera** [comment]
@swails : These failures seem unrelated to the changes I've made. Perhaps travis just needs to be restarted?

**Comment by jchodera** [comment]
Maybe this is a change in some upstream dependency?
```
======================================================================
ERROR: Parsing Amber netcdf file that does not have box info
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ParmEd/ParmEd/test/test_parmed_amber.py", line 1689, in test_parsing_netcdf_file_without_unitcells
    parm = pmd.load_file(tn, fn)
  File "/home/travis/miniconda/envs/myenv/lib/python3.6/site-packages/parmed/formats/registry.py", line 197, in load_file
    return cls.parse(filename, *args, **kwargs)
  File "/home/travis/miniconda/envs/myenv/lib/python3.6/site-packages/parmed/amber/amberformat.py", line 359, in parse
    return LoadParm(filename, *args, **kwargs)
  File "/home/travis/miniconda/envs/myenv/lib/python3.6/site-packages/parmed/amber/readparm.py", line 73, in LoadParm
    f = load_file(xyz)
  File "/home/travis/miniconda/envs/myenv/lib/python3.6/site-packages/parmed/formats/registry.py", line 176, in load_file
    if cls.id_format(filename):
  File "/home/travis/miniconda/envs/myenv/lib/python3.6/site-packages/parmed/amber/netcdffiles.py", line 57, in id_format
    f = NetCDFFile(filename, 'r', mmap=False)
  File "/home/travis/miniconda/envs/myenv/lib/python3.6/site-packages/parmed/utils/netcdf.py", line 286, in __init__
    self._read()
  File "/home/travis/miniconda/envs/myenv/lib/python3.6/site-packages/parmed/utils/netcdf.py", line 598, in _read
    self.__dict__['version_byte'] = fromstring(self.fp.read(1), '>b')[0]
DeprecationWarning: The binary mode of fromstring is deprecated, as it behaves surprisingly on unicode inputs. Use frombuffer instead
```

**Comment by swails** [comment]
> Maybe this is a change in some upstream dependency?

No, that's Python 3.6 deprecating something that `netcdf.py` uses :(

File: parmed/modeller/residue.py [hunk 1]
```diff
@@ -220,6 +220,18 @@ def add_bond(self, atom1, atom2, order=1.0):
...
+        """ Delete a bond from this residue template.
+
+        Parameters
+        ----------
+        bond : :class:`Bond`
+            The bond to be deleted
+
+        """
+        bond.delete()
+        self.bonds.remove(bond)
```
Comment (swails) [inline parmed/modeller/residue.py]: This should be done IFF `bond` is in `self.bonds`, yes?

File: parmed/openmm/parameters.py [hunk 2]
```diff
@@ -91,6 +91,32 @@ def __init__(self, *filenames):
...
+        for atom in residue.atoms:
+            if atom.type not in params.atom_types_str:
+                raise Exception('Residue {} contains atom type {} not found in parameter set'.format(residue.name, atom.type))
+            atom.atomic_number = params.atom_types_str[atom.type].atomic_number
+
+        # Check waters
+        if residue.empirical_chemical_formula == 'H2O':
+            # Remove any H-H bonds if they are present
+            for bond in residue.bonds:
+                if (bond.atom1.element_name == 'H') and (bond.atom2.element_name == 'H'):
```
Comment (swails) [inline parmed/openmm/parameters.py]: Is `element_name` an attribute?  I don't ever remember that being an attribute.  Why not use `atomic_number`?

File: parmed/openmm/parameters.py [hunk 3]
```diff
@@ -91,6 +91,32 @@ def __init__(self, *filenames):
...
+                    LOGGER.debug('Deleting H-H bond from water residue {}'.format(residue.name))
+                    residue.delete_bond(bond)
```
Comment (swails) [inline parmed/openmm/parameters.py]: You can't do this.  `residue.delete_bond` modifies `residue.bonds` as you're iterating over it.  Python doesn't like that.

You either need to loop over `list(residue.bonds)` or keep a list of bonds to remove from `self.bonds` after the iteration over all bonds is done.

**Comment by jchodera** [comment]
Thanks for the great feedback! Working on this now!

Two quick questions for @swails:
* @rafwiewiora notes that the previous behavior was that `OpenMMParameterSet` would happily import a parameter set where some residue templates had missing atom type definitions, and would just silently not write out those residues when exporting an `ffxml`. Are you happy to keep that behavior?
* Should parameter set readers guaranteed to set their `Residue.atom.atomic_num` fields to the correct atomic numbers, or is this something that is not guaranteed to be defined at any time?

File: parmed/openmm/parameters.py (hunk 2 above)
Comment (jchodera) [reply parmed/openmm/parameters.py]: I just added `element_name` as a convenience property, since this makes the code clearer to understand.

**Comment by jchodera** [comment]
I think I've added the necessary test coverage, and have changed the code to drop residues without atom type definitions when converting to `OpenMMParameterSet`. If that's not the desired behavior, I'm happy to fix!

**Comment by swails** [comment]
> @rafwiewiora notes that the previous behavior was that OpenMMParameterSet would happily import a parameter set where some residue templates had missing atom type definitions, and would just silently not write out those residues when exporting an ffxml. Are you happy to keep that behavior?

Yes, I'm happy to keep that behavior.

> Should parameter set readers guaranteed to set their Residue.atom.atomic_num fields to the correct atomic numbers, or is this something that is not guaranteed to be defined at any time?

This isn't something that can really be guaranteed, as not all file formats specify this information.  If a good guess can be made somewhat unambiguously (either through comparing atomic masses or some heuristic combination of various properties, like bond pattern + name), then that guess is made.  Otherwise, it's just set to `None`.

**Comment by swails** [comment]
Is this ready?

**Comment by jchodera** [comment]
Yes!

**Pull request merged**